add strategy paper for observability (#695)
kyma-project/community: concepts/observability-strategy/configurable-logging/01-logging-groundwork.md

As outlined in the [general strategy](../strategy.md), integration (and with that, changing the focus away from in-cluster backends) is the key to open up the stack for a broad range of use cases. Users can simply bring their own backends if they already use a commercial offering or run their own infrastructure. The data can be stored outside the cluster in a managed offering, shared with the data of multiple clusters, away from any tampering or deletion attempt of a hacker, to name just a few.

This concept proposes how to open up to those new scenarios by supporting convenient integration at runtime.

## Requirements

### Basic backend configuration

> - Have a vendor-neutral agent layer that collects and ships the telemetry data, but does not permanently store it (as a backend)

Reviewer: didn't we want to throw out the term "agent layer" ?

Reviewer: To follow up on our brief call this morning: "Agent layer" is not wrong, it's a perfectly fine term. But if we want to use it, we need to define the term, and then it maybe should also show up in the diagrams. 
Alternatively, we can decide that the term isn't need it and just say "collectors", "collector layer" or something like that; sticking to words we're already using.

Author: fixed

# Configurable Logging: Groundwork

## Current Situation and Motivation

![a](./assets/logging-current.drawio.svg)

In the current setup, the logging component provides a feature-rich end-to-end logging solution with Loki as backend and Grafana as log browser. Users can configure Fluent Bit, the log collector, flexibly and neutrally - but only during installation, otherwise the configuration is lost at the next Kyma upgrade. Mixing and matching different integrations is hard, because users must deal with one centralized configuration (the Fluent Bit config) at deployment.
Furthermore, Loki is provided in a non-scalabale setup and cannot be configured at runtime.

> As outlined in the [general strategy](../strategy.md), integration (and with that, changing the focus away from in-cluster backends) is the key to open up the stack for a broad range of use cases. Users can simply bring their own backends if they already use a commercial offering or run their own infrastructure. The data can be stored outside the cluster in a managed offering, shared with the data of multiple clusters, away from any tampering or deletion attempt of a hacker, to name just a few.

Reviewer: ```suggestion
As outlined in the [general strategy](../strategy.md), integration (and with that, changing the focus away from in-cluster backends) is the key to open up the stack for a broad range of use cases. Users can simply bring their own backends if they already use a commercial offering or run their own infrastructure. To name just a few advantages, the data can be stored outside the cluster in a managed offering, shared with the data of multiple clusters, and kept away from any tampering or deletion attempts by hackers.
```

This concept proposes how to open up to those new scenarios by supporting convenient integration at runtime.

## Requirements

### Basic backend configuration
- Have a vendor-neutral layer of collectors that collects and ships the telemetry data, but does not permanently store it (as a backend)
- Support configuration of the selected collector at runtime (no need to run a Kyma upgrade process) in a scenario-focused approach
- Support multiple configurations at the same time in individual resources to enable easy activation of dedicated scenarios

> - As a minimum, support one vendor-neutral input and output. It should be possible to chain your custom collector for specific conversions. For example, for traces, supporting the OLTP protocol will support most of the vendors already. Chaining a custom OpenTelemetry  Collector can do custom conversion to a specific protocol.

Reviewer: ```suggestion
- As a minimum, support one vendor-neutral input and output. It should be possible to chain your custom collector for specific conversions. For example, for traces, supporting the OLTP protocol will support most of the vendors already. Chaining a custom OpenTelemetry Collector can do custom conversion to a specific protocol.
```

- Support all typical settings for the supported outputs of the used collector, do not hide or abstract them
- The collector must run stable at any time. Bad configuration must be prevalidated and rejected. Fast feedback is welcome.
- Secrets must be kept secret.

> - Scenarios/Pipelines need to be isolated and have it's own buffer management. If a backend is in a bad shape and cannot process any data anymore, data should still continue to be pushed to other backends

Reviewer: ```suggestion
- Scenarios and pipelines must be isolated and have their own buffer management. If a backend is in a bad shape and cannot process any data anymore, data should still continue to be pushed to other backends.
```

> - Typical auth mechanisms for the integration must be support, especially also client certificate based solutions

Reviewer: ```suggestion
- Typical authentication mechanisms for the integration must be supported, especially solutions based on client certificates.
```

- For logging: de-dotting for elasticsearch must be possible

> - Filtering of unrelevant data (like dropping logs of kyma-system Namespace) must be possible

Reviewer: ```suggestion
- Filtering of irrelevant data (like dropping logs of kyma-system Namespace) must be possible.
```

### Template definitions
- Have a mechanism to provide templates and best practices for typical scenarios, which can be instantiated at runtime

> - Such template provides the same feature-richness as a configuration scenario

Reviewer: ```suggestion
- Such templates provide the same feature-richness as a configuration scenario.
```

> - If specific templates are bundled with Kyma, updates of a template in use must propagate to the actual collector configuration

Reviewer: ```suggestion
- If specific templates are bundled with Kyma, updates of a template in use must propagate to the actual collector configuration.
```

  - Support placeholder definitions with default values and descriptions

### Template instantiation

> - A template is instantiated by binding it to a secret that provides input for placeholders like URL and credentials. Alternatively, placeholders can be filled out with configMaps and static values.

Reviewer: ```suggestion
- A template is instantiated by binding it to a secret that provides input for placeholders like URL and credentials. Alternatively, placeholders can be filled out with ConfigMaps and static values.
```

>   - The template validates whether all placeholders will be replaced with values.

Reviewer: ```suggestion
  - The template validates whether all placeholders have been replaced with values.
```
It can only validate that after it happened, or am I misunderstanding something here?

> - A template can be instantiated for a specific workload and/or Namespaces only.

Reviewer: ```suggestion
- A template can be instantiated for a specific workload and/or Namespace only.
```

starting with "a specific", this must be singular

### Local backend
- Kyma will provide a blueprint based on Helm for installing the typical Loki stack as example.
- The setup is not meant to be HA and scalable.

> ## Proposed Solution

Reviewer: ```suggestion
## Proposed solution
```

Sentence case in titles ¯\_(ツ)_/¯

# Configurable Logging: Groundwork

## Current Situation and Motivation

![a](./assets/logging-current.drawio.svg)

> In the current setup, all three observability aspects (log, trace, metric) provide a preconfigured backend with visualisations. However, they don't provide a neutral and unified way to integrate backends outside of the cluster.

Reviewer: Why do we talk about all three observability aspects (and in line 9, about tracing specifically) in the logging-groundwork document?
The monitoring-groundwork doc focuses on monitoring only.

![a](./assets/logging-current.drawio.svg)

> In the current setup, the logging component provides feature-rich e2e logging solution with Loki as backend and Grafana as log Browser. The used log collector (FluentBit) can be configured flexibly and neutrally. However, users must apply the configuration during installation; otherwise it's lost at the next Kyma upgrade process. Furthermore, it is hard to mix and match different integrations, as you must deal with one centralized configuration (the Fluent Bit config) at deploy time. Furthermore, Loki is provided in a non-scalabale setup and cannot configured at runtime.

Reviewer: ```suggestion
In the current setup, the logging component provides a feature-rich end-to-end logging solution with Loki as backend and Grafana as log browser. Users can configure Fluent Bit, the log collector, flexibly and neutrally - but only during installation; otherwise it's lost at the next Kyma upgrade. Mixing and matching different integrations is hard, because you must deal with one centralized configuration (the Fluent Bit config) at deployment.
Furthermore, Kyma provides Loki in a non-scalabale setup, which users cannot configured at runtime.
```
